### ISSUE TITLE
Supporting count-distinct

### ISSUE THREAD

**Issue by pyongjoo** [mozafari/verdictdb#242]
We should exploit the property of HyperLogLog.

**Comment by pyongjoo** [comment]
@Beastjoe Can you see if
1. we can make no AsyncAgg is done for count-distinct
2. if the backend database supports approximate count-distinct, we use those native functions in the sql writing stage (SqlSyntax class may be modified somehow). FYI, Redshift, Impala certainly have them; Spark seemed to have one too, but please double-check 

In the future, we can implement our own HLL algorithm as UDF for Hive-like engines. Our UDF should return hash values before averaging, so we can average them on VerdictDB side. The major exception is Redshift, which does not provide an efficient way to define UDF. I read online that Its Python-based UDF is much slower.

**Comment by Beastjoe** [comment]
I am thinking the case when both count-distinct and other aggregate function show up in a query that contain scramble table. Is it appropriate to put count-distinct only in the first individual aggregate node and pass it to the AsyncAggNode?

**Comment by pyongjoo** [comment]
I think that approach will make the count-distinct incorrect. Are they test queries that include both count-distinct and other aggregates in a single select list?

**Comment by Beastjoe** [comment]
Yes, I notice that this method is not good. Test queries do not have this case.

**Comment by pyongjoo** [comment]
Then, what we can do is simply reject the queries including both by including error checks. I suggest placing an additional step (as a separate function) after query standardization to check if the submitted query satisfies the structure we can support.
